### ISSUE TITLE
would like to alter max number of epochs while training loop is running

### ISSUE THREAD

**Issue by sneiman** [PyTorchLightning/pytorch-lightning#762]
I would like to be able to change the maximum number of epochs using a call added to trainer.

In prior work, I have used a run time monitor that allows the user to run self defined commands during training. One major benefit is to end a training run early while still doing testing, or to extend a training run that is going well but could use more epochs. Other things have proven useful - including setting flags in the model, changing rates, etc.

The lightning training loop constructs a range using trainer.max_epochs. If this loop was implemented as an index and test, altering self.max_epochs would have the desired effect. This should be done with a call to trainer(trainer.change_max_epochs()?) so that error checking can be performed.

One might object that getting user input during training has some serious problems. The challenging part of doing this is to poll stdin so that the various loops do not pause waiting for input. I have this for Ubuntu, and have it working well with both dp and ddp, and believe it is similarly straightforward for other linuxes, MacOS, and Windows.

Another objection might be that one can always resubmit the model for additional training. Certainly true, but I frequently check training status as I'm headed to bed and realize that I would like to extend the current session another 50 or 100 epochs while it is working overnight.

Not a burning issue. but a nice to have. I do have an architecture that accomplishes all this, and am happy to adapt and contribute it if you think it worthwhile.

**Comment by williamFalcon** [comment]
But if  the program is running, how exactly would you do that? On a  notebook you could stop the current cell and call .fit() again on a different cell to continue

**Comment by sneiman** [comment]
I think it's pretty straightforward, unless I am missing something. If we are only changing epoch count, just need to substitute a loop with exposed variables instead of the for ... range() call. There are no synch issues as the impact of changing max_epoch isn't seen until the top of the while loop.

Have to expose the index variable, and keep it in trainer object so it is accessible to the set predicate. Changing it would require `set_max_epochs()` call to test that the new value is at least 1 more than the current. At each loop, a client would call 'get_max_epochs' to get the latest value. This should happen immediately after the first sync point at the end of an epoch to keep all subprocesses in sync.

Am I missing something? I haven't looked hard at where the sync points are, and perhaps this would be prohibited on multi node jobs - unless they are synced, too.

**Comment by sneiman** [comment]
BTW, I have no experience with Jupyter and similar, so cannot comment there. Please enlighten me if I have missed a critical issue.

**Comment by chutaklee** [comment]
This package may meet some of your requests. https://github.com/julvo/reloading

**Comment by Borda** [comment]
I think that the simplest way is to use [EarlyStopping](https://pytorch-lightning.readthedocs.io/en/latest/early_stopping.html?highlight=earlystop) or implement [Custom Callback](https://pytorch-lightning.readthedocs.io/en/latest/callbacks.html#callbacks) which may change the `max_epochs` according to given rule...
Thx you for your interest and interesting point, feel free to reopen if needed :robot:

**Comment by sneiman** [comment]
Makes sense to close - thanks. I have a running version on my local fork. Waiting til after the scheduled refactoring to work it up completely and submit a pr.

Doing with callbacks() does work, but the loops also have to change - range()'s don't change after they are created. Also, a shared value is needed to get the new value to all the processes. And I believe an mp Manager() is needed to move between nodes - though I haven't tried this yet. That's what is waiting for the refactoring.

The tricky part is letting the user interact via the console to interrupt things and reset the value. Needs polling and some slightly more involved console interaction. But also need to know **where is the console in a multi-node job**? I have built what is needed for multi processing on one node, and use it every day. But need a good working assumption as to where the console is on a multi-node job.

FWIW, this is also useful to adjust other things - learning rates, weight decay, etc. I have an initial extensible arch to do all of that.

Thanks for all you guys do. This is a hugely valuable project. **And congrats on your funding - I ran one of the larger VC firms for a long time - am happy to be an independent sounding board for you guys if you ever need it.**
